ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs

```diff
@@ -45,9 +52,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
 import           Ouroboros.Consensus.HardFork.Combinator.Mempool
 import           Ouroboros.Consensus.HardFork.Combinator.Serialisation.Common
 import           Ouroboros.Consensus.HardFork.Combinator.Serialisation.SerialiseDisk ()
+import qualified Ouroboros.Consensus.HardFork.History as History
+
+{-------------------------------------------------------------------------------
+  Serialisation of products
+-------------------------------------------------------------------------------}
+
+-- | Encoding of `NP f xs` while filtering out the components for disabled eras
+-- as indicated by the `HardForkNodeToClientVersion xs`. disabled eras imply
+-- that the protocol version does not support those eras. Hence, omitting the
```

Comment:
> disabled eras imply that the protocol version does not support those eras. Hence, omitting the corresponding elements is the correct behavior.

Are we assuming that the eras are in a sequence and that some prefix is enabled while the complementary suffix is disabled? We should `assert` that shape, if we are.

Reply:
That is indeed a property of the existing node-to-client versions and I expect the same from future versions. This function does **not** AFAICS make use of that assumption: we simple omit disabled eras regardless of position their. Hence I don't see a reason to add an assertion here. Am I missing something?

Reply:
In your mind what is an "expected arrangement"? Do you mean a run of 0 or more enabled followed by a run of 0 or more disabled? If we really want to enforce that property, then maybe we should change the `HardForkNodeToClientEnabled` constructor to store a `Telescope EraNodeToClientVersion EraNodeToClientVersion xs` instead of a `NP EraNodeToClientVersion xs` or something like that?

For now I've just added an assertion here and in `decodeNodeToClientNP`.

Reply:
Yep, that's what I meant. Interesting idea to use a `Telescope`.

I've opened https://github.com/input-output-hk/ouroboros-network/issues/3303

Thanks. Resolving this convo.